### PR TITLE
Add StdNames

### PR DESCRIPTION
**Pull request by ceferisbarov**
I am still trying to make the code more readable. Suggestions are welcome.

### CODE REVIEW

File: src/transforms/stdnames.jl [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    StdNames(S)
+
+StdNames transform standardizes the column names.
+"""
+struct StdNames <: Stateless
+    stdmethod::Symbol
```
Comment (juliohm) [inline src/transforms/stdnames.jl:11]: We use 2-space indentation

Comment (juliohm) [inline src/transforms/stdnames.jl:11]: rename the field to `spec`

File: src/transforms/stdnames.jl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+isrevertible(::Type{StdNames}) = true
+
+function apply(transform::StdNames, table)    
+    if transform.stdmethod == :camel
+        _camel(table)
+    elseif transform.stdmethod == :snake
+        _snake(table)
+    else
+        _upper(table)
+    end
```
Comment (juliohm) [inline src/transforms/stdnames.jl:23]: You can use the short cut:

```julia
spec = transform.spec

spec == :camel && _camel(table)
spec == :snake && _snake(table)
spec == :upper && _upper(table)
```

File: src/transforms/stdnames.jl [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+end
+
+function revert(::StdNames, newtable, cache)
+    names = Tables.columnnames(newtable)
+    namesdict = Dict(zip(names, cache))
+    newtable |> Rename(namesdict)
```
Comment (juliohm) [inline src/transforms/stdnames.jl:29]: Try to use the same variable names used in other transforms.

File: src/transforms/stdnames.jl [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    table |> Rename(namesdict), oldnames
+end
+
+function _upper(table)
+    oldnames = Tables.columnnames(table)
+    upperfunction(x) = Symbol(uppercase(string(x)))    
+    newnames = map(upperfunction, oldnames)
+    namesdict = Dict(zip(oldnames, newnames))
+    table |> Rename(namesdict), oldnames
+end
```
Comment (juliohm) [inline src/transforms/stdnames.jl:59]: Notice how these three functions are repeating the same code and logic. Try to move the common pieces to the `apply` function and then only call specific _camel(str) on the names. In other words, your `apply` should contain all logic and loop over all names performing the standardization with one of three functions _camel, _snake and _upper, which should operate on strings.

File: test/transforms.jl [hunk 5]
```diff
@@ -680,6 +680,34 @@
     @test n1 == n2
   end
 
+  @testset "StdNames" begin
```
Comment (juliohm) [inline test/transforms.jl:683]: Make sure you include tests with row tables. Take a look at other transforms for examples.

**Review by juliohm — CHANGES_REQUESTED** [review]
Great start @ceferisbarov , left a first round of reviews.

File: src/transforms/stdnames.jl (hunk 2 above)
Comment (ceferisbarov) [reply src/transforms/stdnames.jl:23]: This doesn't work as expected. It works fine while `spec = :upper`, but it returns `false` otherwise.

Comment (juliohm) [reply src/transforms/stdnames.jl:23]: I may have missed parenthesis. What about `(spec == :camel) && _camel(table)`? Anyways, I think you will need to refactor the logic a bit more to move more lines of code to apply and refactor _camel, _snake and _upper to take strings as input.

File: test/transforms.jl [hunk 6]
```diff
@@ -680,6 +680,34 @@
...
+    a = [3, 2, 1, 4, 5, 3]
+    b = [2, 4, 4, 5, 8, 5]
+    t = Table(; a, b)
+    t = t |> Rename(:a => "apple tree", :b => "banana tree")
```
Comment (juliohm) [inline test/transforms.jl:687]: You can create a table with the desired column names instead of using Rename here.

Comment (juliohm) [reply test/transforms.jl:687]: Also, in this example make sure that the table also contains clean column names such as `a`, `B`. They shouldn't be affected by some of the conventions.

Comment (ceferisbarov) [reply test/transforms.jl:687]: @juliohm  I have tried using a Symbol or a string instead of a variable name, but both of them throw syntax error. I am not sure how I can create a table with these column names.

File: src/transforms/stdnames.jl (hunk 3 above)
Comment (ceferisbarov) [reply src/transforms/stdnames.jl]: I assumed you were referring to `namesdict` variable.

File: src/transforms/stdnames.jl [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+function apply(transform::StdNames, table)  
+  oldnames = Tables.columnnames(table)
+  spec = transform.spec
+
+  (spec == :camel) && (newnames = map(_camel, oldnames))
+  (spec == :snake) && (newnames = map(_snake, oldnames))
+  (spec == :upper) && (newnames = map(_upper, oldnames))
+
+  names = Dict(zip(oldnames, newnames))
+  table |> Rename(names), oldnames
```
Comment (juliohm) [inline src/transforms/stdnames.jl:27]: Now that is 🥇 👏🏽

File: src/transforms/stdnames.jl [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+end
+
+function _camel(name)
+  substrings = split(string(name))
+  capitalize(s) = String([i == 1 ? uppercase(c) : c for (i, c) in enumerate(s)])
+  Symbol(join(map(capitalize, substrings)))
+end
+
+function _snake(name)
+  Symbol(join(split(string(name)), "_"))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:43]: You also need to change everything to `lowercase`

File: src/transforms/stdnames.jl [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+end
+
+function _upper(name)
+  Symbol(uppercase(string(name)))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:47]: You also need to remove whitespace

File: test/transforms.jl [hunk 10]
```diff
@@ -680,6 +680,49 @@
     @test n1 == n2
   end
 
+  @testset "StdNames" begin
+    names = (:a, Symbol("apple tree"), Symbol("banana tree"))
```
Comment (juliohm) [inline test/transforms.jl:684]: I suggest adding more complicated names such as `foo_Bar`, `A b`, ... you need to test the names extensively.

Ideally you will introduce a small table for each kind of name issue. For example, you can introduce a table with names that have underscore, another table with names that have spaces, another table with upper and lower case mixed, etc. This will cover more issues.

**Review by juliohm — CHANGES_REQUESTED** [review]

File: src/transforms/stdnames.jl (hunk 7 above)
Comment (juliohm) [reply src/transforms/stdnames.jl:27]: We can do even better here and save the Rename transform as the cache, so that the revert is just a single call to revert with the cache transform.

File: src/transforms/stdnames.jl [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+struct StdNames <: Stateless
+  spec::Symbol
+end
+
+StdNames() = StdNames(:upper)
+
+isrevertible(::Type{StdNames}) = true
+
+function apply(transform::StdNames, table)  
+  oldnames = Tables.columnnames(table)
```
Comment (eliascarv) [inline src/transforms/stdnames.jl:19]:
```suggestion
  cols = Tables.columns(table)
  oldnames = Tables.columnnames(cols)
```
This is the correct way to use the `Tables.columnnames` function.

**Review by eliascarv — CHANGES_REQUESTED** [review]

File: src/transforms/stdnames.jl (hunk 11 above)
Comment (ceferisbarov) [reply src/transforms/stdnames.jl:19]: Thanks!

File: src/transforms/stdnames.jl (hunk 7 above)
Comment (ceferisbarov) [reply src/transforms/stdnames.jl:27]: I did this, but it looks a little bit off. `revert` function of the `Rename` transform takes three arguments `(transform, table, cache)`, but only uses two `(transform, table)`. So we get something like this:
```
revert(cache, newtable, cache)
```
The first cache is cache of StdNames transform. The second cache is supposes to be cache of the Rename transform, but it is not used inside the function anyway, so I just wrote it like this. Wait until I push the commit. It will be cleared then.

Comment (eliascarv) [reply src/transforms/stdnames.jl:27]: I think you are not understanding. This is what Júlio is trying to say:
```julia
function apply(transform::StdNames, table)
  # code...
  names = Dict(zip(oldnames, newnames))
  rtrans = Revert(names)
  newtable, cache = apply(rtrans, table)
  newtable, (rtrans, rcache)
end

revert(::StdNames, newtable, cache)
  rtrans, rcache = cache
  revert(rtrans, newtable, rcache)
end
```

Comment (eliascarv) [reply src/transforms/stdnames.jl:27]: I think you didn't understand what Júlio meant. This would be the code:
```julia
function apply(transform::StdNames, table)
  # code...
  names = Dict(zip(oldnames, newnames))
  rtrans = Revert(names)
  newtable, cache = apply(rtrans, table)
  newtable, (rtrans, rcache)
end

revert(::StdNames, newtable, cache)
  rtrans, rcache = cache
  revert(rtrans, newtable, rcache)
end
```

Comment (ceferisbarov) [reply src/transforms/stdnames.jl:27]: Oh, I see. Thanks again. You are doing most of the work for me here :))

File: test/transforms.jl (hunk 10 above)
Comment (ceferisbarov) [reply test/transforms.jl:684]: @juliohm  I know we already discussed this, but don't you think we should also split at underscores? I actually think it would be a reasonable thing to do.

**Comment by ceferisbarov** [comment]
@juliohm  Test set grew too large. I think we can create a function to avoid repeating the same lines.

File: test/transforms.jl (hunk 10 above)
Comment (juliohm) [reply test/transforms.jl:684]: @ceferisbarov it depends on the transform, for `:snake` we don't want to split underscores.

**Review by juliohm — CHANGES_REQUESTED** [review]
That is looking great now @ceferisbarov. I have a final suggestion before we merge this. Can you please replace all tests you have written by simple tests of the form:

```julia
@test TableTransforms._camel("apple banana") == "AppleBanana"
@test TableTransforms._snake("apple banana") == "apple_banana"
...
```

That way we can make sure that we cover various types of strings.

After the above set of tests you can simply add two more tests with tables to make sure that `apply` and `revert` are working with column and row tables. Also `reapply` and `isrevertible`.

File: src/transforms/stdnames.jl [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  substrings = split(string(name))
+  Symbol(join(map(uppercasefirst, substrings)))
+end
+
+function _snake(name)
+  Symbol(lowercase(join(split(string(name)), "_")))
+end
+
+function _upper(name)
+  Symbol(replace(uppercase(string(name)), " "=>""))
```
Comment (eliascarv) [inline src/transforms/stdnames.jl:48]:
```suggestion
  Symbol(replace(uppercase(string(name)), " " => ""))
```
Just a small code style change.

File: src/transforms/stdnames.jl [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  Symbol(join(map(uppercasefirst, substrings)))
+end
+
+function _snake(name)
+  Symbol(lowercase(join(split(string(name)), "_")))
+end
+
+function _upper(name)
+  Symbol(replace(uppercase(string(name)), " " => ""))
+end
```
Comment (juliohm) [inline src/transforms/stdnames.jl:49]: Refactor these functions so that they receive String and output String. Then in the main `apply` call you can convert from and to Symbol. This will clean up the code considerably.

File: src/transforms/stdnames.jl [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+isrevertible(::Type{StdNames}) = true
+
+function apply(transform::StdNames, table)  
+  cols = Tables.columns(table)
+  oldnames = Tables.columnnames(cols)
+  spec = transform.spec
+
+  (spec == :camel) && (newnames = map(_camel, map(string, oldnames)))
+  (spec == :snake) && (newnames = map(_snake, map(string, oldnames)))
+  (spec == :upper) && (newnames = map(_upper, map(string, oldnames)))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:25]: Instead of using map(f, xs) everywhere, try to use the broadcast syntax: `string.(oldnames)`

File: src/transforms/stdnames.jl [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+  newnames = map(Symbol, newnames)
```
Comment (juliohm) [inline src/transforms/stdnames.jl:27]: You don't overwrite variables in Julia because this will lead to type instability often.

File: src/transforms/stdnames.jl [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  names = Dict(zip(oldnames, newnames))
+  rtrans = Rename(names)
+  newtable, rcache = apply(rtrans, table)
+  newtable, (rtrans, rcache)
```
Comment (juliohm) [inline src/transforms/stdnames.jl:31]: Notice that this paragraph contains different steps of the algorithm. Try to split lines into paragraphs that make a cohesive sense.

File: src/transforms/stdnames.jl [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  join(map(uppercasefirst, substrings))
+end
+
+function _snake(name)
+  lowercase(join(split(name), "_"))
+end
+
+function _upper(name)
+  replace(uppercase(name), " " => "")
+end
```
Comment (juliohm) [inline src/transforms/stdnames.jl:50]: Note the inconsistency. Sometimes you create an auxiliary variable substrings, sometimes you don't. Try to improve your consistency writing code over time. Also, these functions could be oneliners.

Comment (juliohm) [reply src/transforms/stdnames.jl:50]: Replace the `map` by dot syntax when possible.

File: test/transforms.jl [hunk 18]
```diff
@@ -680,6 +680,49 @@
     @test n1 == n2
   end
 
+  @testset "StdNames" begin
+    # camel, snake, upper tests
+    @test TableTransforms._camel("apple banana") == "AppleBanana"
+    @test TableTransforms._snake("apple banana") == "apple_banana"
+    @test TableTransforms._upper("apple banana") == "APPLEBANANA"
+
+    @test TableTransforms._camel("apple_banana") == "Apple_banana"
```
Comment (juliohm) [inline test/transforms.jl:689]: The camel case should remove the underscore and handle the second word with a first capital letter.

File: test/transforms.jl [hunk 19]
```diff
@@ -680,6 +680,49 @@
...
+    @test TableTransforms._snake("apple_banana") == "apple_banana"
+    @test TableTransforms._upper("apple_banana") == "APPLE_BANANA"
```
Comment (juliohm) [inline test/transforms.jl:691]: I think the upper case should also remove the underscore?

File: test/transforms.jl [hunk 20]
```diff
@@ -680,6 +680,49 @@
...
+
+    @test TableTransforms._camel("apple_Banana") == "Apple_Banana"
+    @test TableTransforms._snake("apple_Banana") == "apple_banana"
+    @test TableTransforms._upper("apple_Banana") == "APPLE_BANANA"
```
Comment (juliohm) [inline test/transforms.jl:695]: I think we should organize these tests differently. For example, consider the camel case. Your would like to write a list of strings like `["apple banana", "Apple_banana", ...]` and then apply the camel case function to all elements of the list to make sure all results are equal to `"AppleBanana"`.

File: test/transforms.jl [hunk 21]
```diff
@@ -680,6 +680,49 @@
...
+    @test isrevertible(T)
+    rtₒ = revert(T, n, c)
+    @test rt == rtₒ
+
+    # reapply test
+    T = StdNames()
+    n1, c1 = apply(T, rt)
+    n2 = reapply(T, n1, c1)
+    @test n1 == n2
+
```
Comment (juliohm) [inline test/transforms.jl:723]:
```suggestion
```

File: test/transforms.jl [hunk 22]
```diff
@@ -680,6 +680,49 @@
...
+    @test TableTransforms._camel("B") == "B"
+    @test TableTransforms._snake("B") == "b"
+    @test TableTransforms._upper("B") == "B"
+
+    # row table test
+    names = (:a, Symbol("apple tree"), Symbol("banana tree"))
+    cols = ([1,2,3], [4,5,6], [7,8,9])
+    t = Table(; zip(names, cols)...)
+
```
Comment (juliohm) [inline test/transforms.jl:709]: Why there is an empty line here? Can you please try to be more consistent with the way you group lines of code? It helps a lot the reviewer understand your intention.

**Review by juliohm — CHANGES_REQUESTED** [review]
@ceferisbarov please request a review whenever you feel that the PR is ready. Attached you can find another round of suggestions.

File: test/transforms.jl (hunk 20 above)
Comment (ceferisbarov) [reply test/transforms.jl:695]: @juliohm I am trying something like this now, but I came across an edge case. What should `TableTransforms._snake("_A")` return? Do we strip spaces and underscores or not?

**Review by juliohm — CHANGES_REQUESTED** [review]

File: src/transforms/stdnames.jl [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+function apply(transform::StdNames, table)  
+  cols = Tables.columns(table)
+  oldnames = Tables.columnnames(cols)
+  spec = transform.spec
+
+  (spec == :camel) && (newstringnames = _camel.(string.(oldnames)))
+  (spec == :snake) && (newstringnames = _snake.(string.(oldnames)))
+  (spec == :upper) && (newstringnames = _upper.(string.(oldnames)))
+
+  newnames = Symbol.(newstringnames)
```
[the remark this answers is not in the document]
Comment (ceferisbarov) [reply src/transforms/stdnames.jl:28]: Rename struct is defined as:
```
struct Rename <: Stateless
  names::Dict{Symbol,Symbol}
end
```
And its constructors:
```
Rename(names::Pair) = _pairsyms(names) |> Dict |> Rename
Rename(names...) = _pairsyms.(names) |> Dict |> Rename

_pairsyms(x::Pair) = Symbol(first(x)) => Symbol(last(x))
```
We can pass `string => Symbol`, `Symbol => string`, and `string => string` pairs to `Rename`, and it will work. But if we pass them as a dictionary (which is what we are doing in this case), it will throw `MethodError: Cannot convert an object of type String to an object of type Symbol`.

I think this is inconsistent behavior and should be changed. We can update the `Rename` transform to accept strings as part of a dictionary, too. But that is beyond the scope of the PR. @juliohm

**Comment by ceferisbarov** [comment]
@juliohm  We have another interesting edge case. Different column names can become same after transformation. For example, `apple_tree` and `appletree` would both be transformed into `APPLETREE` via `StdNames(:upper)` transform. This throws the following error:
```
duplicate field name in NamedTuple: "APPLETREES" is not unique
```
Now it obviously doesn't make sense to name two different columns `apple_tree` and `appletree`, but it is possible in theory. What do you suggest?

**Comment by ceferisbarov** [comment]
@julai Sorry, I read your last comment only after writing mine :)

File: src/transforms/stdnames.jl (hunk 23 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/transforms/stdnames.jl:28]: Yes, we can work on that on a separate PR.

**Comment by juliohm** [comment]
@ceferisbarov I think I already suggested a fix for this non-uniqueness issue. Did you read my comment above? You should apply the _camel or whatever function to all column names and then make them unique afterwards as a post-processing step before calling Rename. This post-processing step will append as many _ as necessary to make the columns unique. For example, if the names are A A A they will become A A_ A__

**Comment by juliohm** [comment]
You pinged a different github account

File: src/transforms/stdnames.jl [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+  uniquenames = [names...]
+  for i in range(2,length(uniquenames))
+      prev = uniquenames[1:i-1]
+      while uniquenames[i] in prev
+        uniquenames[i] = string(uniquenames[i], "_")
+      end
+  end
+
+  Tuple(uniquenames)
+end
```
Comment (juliohm) [inline src/transforms/stdnames.jl:55]: Again, you are mixing 2-space indentation with 4-space indentation. Please use an IDE and set this to avoid bumping into this issue again? Also, the implementation is not great. Simply create an empty vector of Symbol and keep pushing to it if the entry is not already in the vector. Don't use indexing.

File: src/transforms/stdnames.jl [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+  
+  spec = transform.spec
+
+  spec == :camel && (newstringnames = _camel.(string.(oldnames)))
+  spec == :snake && (newstringnames = _snake.(string.(oldnames)))
+  spec == :upper && (newstringnames = _upper.(string.(oldnames)))
+
+  newuniquenames = _unique(newstringnames)
+  newnames = Symbol.(newuniquenames)
+  names = Dict(zip(oldnames, newnames))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:30]: Let's first work on a PR to add a Rename with Dict of String, then we can come back here and avoid all this back and forth between Symbol and String.

**Review by juliohm — CHANGES_REQUESTED** [review]
@ceferisbarov please start a separate PR adding a Rename with Dict of Strings. After that we can come back to this PR and finish it.

**Comment by juliohm** [comment]
Now that #71 has been merged, we can come back to this PR. @ceferisbarov please rebase this PR to the latest master branch so that we can use the latest Dict constructor you added 👍🏽

File: src/transforms/stdnames.jl (hunk 24 above)
Comment (ceferisbarov) [reply src/transforms/stdnames.jl:55]: Is this what you had in mind? I don't think our vector needs to be `Vector{Symbol}`, as we are working with strings.
```julia
function unique(names)
  newnames = Vector{String}()
  for name in names
    updatedname = name
    while updatedname ∈ newnames
      updatedname = string(updatedname, "_")
    end
    push!(newnames, updatedname)
  end

  newnames
end
```

File: Project.toml [hunk 26]
```diff
@@ -1,7 +1,7 @@
 name = "TableTransforms"
 uuid = "0d432bfd-3ee1-4ac1-886a-39f05cc69a3e"
 authors = ["Júlio Hoffimann <julio.hoffimann@gmail.com> and contributors"]
-version = "0.5.2"
+version = "1.0.0"
```
Comment (juliohm) [inline Project.toml:4]: This change should not show up in this PR. I think you rebased with the incorrect commit. Rebasing is always a nightmare. Please feel free to start a fresh PR if you can't sort out the rebase.

**Review by juliohm — CHANGES_REQUESTED** [review]

File: src/transforms/stdnames.jl [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+  spec = transform.spec
+
+  spec == :camel && (nonuniquenames = _camel.(oldnames))
+  spec == :snake && (nonuniquenames = _snake.(oldnames))
+  spec == :upper && (nonuniquenames = _upper.(oldnames))
+
+  newnames = _unique(nonuniquenames)
+  names = Dict(zip(oldnames, newnames))
+
+  rtrans = Rename(names)
```
Comment (juliohm) [inline src/transforms/stdnames.jl:31]:
```suggestion
  spec == :camel && (names = _camel.(oldnames))
  spec == :snake && (names = _snake.(oldnames))
  spec == :upper && (names = _upper.(oldnames))

  newnames = _unique(names)
  oldnew = Dict(zip(oldnames, newnames))

  rtrans = Rename(oldnew)
```

File: src/transforms/stdnames.jl [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+function revert(::StdNames, newtable, cache)
+  rtrans, rcache = cache
+  revert(rtrans, newtable, rcache)
+end
+
+const delim = ['_', ' ']
+
+function _unique(names)
+  newnames = Vector{String}()
```
Comment (juliohm) [inline src/transforms/stdnames.jl:45]:
```suggestion
  newnames = String[]
```

File: src/transforms/stdnames.jl [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+  for name in names
+    updatedname = name
+    while updatedname ∈ newnames
+      updatedname = string(updatedname, "_")
+    end
+    push!(newnames, updatedname)
+  end
```
Comment (juliohm) [inline src/transforms/stdnames.jl:52]:
```suggestion
  for name in names
    n = name
    while n ∈ newnames
      n = string(n, "_")
    end
    push!(newnames, n)
  end
```

File: src/transforms/stdnames.jl [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+
+  newnames
+end
+
+_camel(name) = join(uppercasefirst.(split(strip(name, delim), delim)))
+
+_snake(name) = lowercase(join(split(strip(name, delim), delim), '_'))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:59]: Can you follow the same pattern of the _camel function? That is, use join at the end after the lowercase is applied to every word.

File: src/transforms/stdnames.jl [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+
+_upper(name) = replace(uppercase(strip(name, delim)), delim => "")
```
Comment (juliohm) [inline src/transforms/stdnames.jl:61]: Do we really need the `strip` when we have `replace`?

File: test/transforms.jl [hunk 32]
```diff
@@ -694,6 +694,68 @@
     @test n1 == n2
   end
 
+  @testset "StdNames" begin
+    # camel test
```
Comment (juliohm) [inline test/transforms.jl:698]: This is not the camel test, please update or remove the comment.

File: test/transforms.jl [hunk 33]
```diff
@@ -694,6 +694,68 @@
...
+    end
+
+    names = ["a", "A", "_a", "_A", "a ", "A "]
+    for name in names
+      @test TableTransforms._camel(name) == "A"
+      @test TableTransforms._snake(name) == "a"
+      @test TableTransforms._upper(name) == "A"
+    end
+
+    # invariant test
```
Comment (juliohm) [inline test/transforms.jl:713]:
```suggestion
    # invariance test
```

**Review by juliohm — CHANGES_REQUESTED** [review]

File: src/transforms/stdnames.jl [hunk 34]
```diff
@@ -0,0 +1,61 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    StdNames(:spec)
+
+StdNames transform standardizes the column names.
```
Comment (juliohm) [inline src/transforms/stdnames.jl:8]: Please improve the docstring and write a text that helps end users understand the options. Try to follow the format of other docstrings.

**Comment by juliohm** [comment]
@ceferisbarov please add the transform to the docs as usual.

File: src/transforms/stdnames.jl [hunk 35]
```diff
@@ -0,0 +1,69 @@
...
+    while n ∈ newnames
+      n = string(n, "_")
+    end
+    push!(newnames, n)
+  end
+
+  newnames
+end
+
+_camel(name) = join(uppercasefirst.(split(strip(name, delim), delim)))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:65]: Do we really the strip here?

**Review by juliohm — CHANGES_REQUESTED** [review]
Can you also add tests for columns with special characters such as `A-B`, `A#`, etc?

File: src/transforms/stdnames.jl (hunk 35 above)
Comment (ceferisbarov) [reply src/transforms/stdnames.jl:69]: We need it for snake case, otherwise, for example, "_A" gets split into "" and "A" and when we combine them, we get "_A" again.
  
Not sure why added this for camel case, TBH.

File: src/transforms/stdnames.jl [hunk 36]
```diff
@@ -0,0 +1,69 @@
...
+    StdNames(:spec)
+
+The transform that standardizes the column names. There are three different methods of standardization: uppercase, camelcase, and snakecase.
+They can be specified by the `spec`` argument (`:upper`, `:camel`, and `:snake`, respectively). Transform defaults to uppercase. 
+
+# Examples
+
+```julia
+StdNames()
+StdNames(:camel)
```
[the remark this answers is not in the document]
Comment (ceferisbarov) [reply src/transforms/stdnames.jl]: I tried something with tables.

File: src/transforms/stdnames.jl [hunk 37]
```diff
@@ -0,0 +1,73 @@
...
+
+# Examples
+
+| Transform | Original Col. Name | Transformed Col. Name |
+| ----------- | ----------- |	----------- |
+| `StdNames`    		 | apple_Trees | APPLETREES |
+| `StdNames(:upper)` | apple_Trees | APPLETREES |
+| `StdNames(:camel)` | apple_Trees | AppleTrees |
+| `StdNames(:snake)` | apple_Trees | apple_trees |
+
```
Comment (juliohm) [inline src/transforms/stdnames.jl:20]:
```suggestion
    StdNames(spec)

Standardizes column names according to given `spec`.
Default to `:upper` case specification.

# Specs

* `:upper` - Uppercase, e.g. COLUMNNAME
* `:camel` - Camelcase, e.g. ColumnName
* `:snake` - Snakecase, e.g. column_name
```

File: src/transforms/stdnames.jl [hunk 38]
```diff
@@ -0,0 +1,73 @@
...
+  newtable, (rtrans, rcache)
+end
+
+function revert(::StdNames, newtable, cache)
+  rtrans, rcache = cache
+  revert(rtrans, newtable, rcache)
+end
+
+const delim = ['_', ' ']
```
Comment (juliohm) [inline src/transforms/stdnames.jl:54]: Can you please move this variable to near where it is used? Right before the _camel function. Please try to think like the reader of the code always.

Comment (juliohm) [reply src/transforms/stdnames.jl:54]: Also, we should probably add `-` to the list of delimiters. So that `day-of-month` gets converted into `DayOfMonth`

File: test/transforms.jl [hunk 39]
```diff
@@ -694,6 +694,84 @@
...
+
+    name = "apple-tree"
+    @test TableTransforms._camel(name) == "Apple-tree"
+    @test TableTransforms._snake(name) == "apple-tree"
+    @test TableTransforms._upper(name) == "APPLE-TREE"
+    
+    name = "apple#"
+    @test TableTransforms._camel(name) == "Apple#"
+    @test TableTransforms._snake(name) == "apple#"
+    @test TableTransforms._upper(name) == "APPLE#"
```
Comment (juliohm) [inline test/transforms.jl:727]: I think we should erase all special characters (except the `delim` characters) in a pre-processing step inside the `apply` function. @eliascarv do you also think this is a good idea?

I think we should set `delim = [' ', '\t', '-', '_']` as our possible delimiters. All other characters could be erased.

Comment (ceferisbarov) [reply test/transforms.jl:727]: I would also suggest adding `&` to the list of delimiters. That being said, I don't think we can make this transform work well with complex column names, anyway. Maybe we can later add more specific transforms that deal with column names (such as `Replace`, but for column names). @juliohm @eliascarv

**Comment by juliohm** [comment]
As I mentioned it should be easy to erase all special symbols as a pre
processing step with a Regex for example. We just need to decide the valids
delimiters. I don't think & is widely used in statistics as a delimiter. It
is used as an interaction term.

Em qua., 11 de mai. de 2022 05:46, Jafar Isbarov ***@***.***>
escreveu:

File: src/transforms/stdnames.jl [hunk 40]
```diff
@@ -0,0 +1,76 @@
...
+* `:snake` - Snakecase, e.g. column_name
+"""
+struct StdNames <: Stateless
+  spec::Symbol
+end
+
+StdNames() = StdNames(:upper)
+
+isrevertible(::Type{StdNames}) = true
+
```
Comment (juliohm) [inline src/transforms/stdnames.jl:24]:
```suggestion
```

File: src/transforms/stdnames.jl [hunk 41]
```diff
@@ -0,0 +1,76 @@
...
+function revert(::StdNames, newtable, cache)
+  rtrans, rcache = cache
+  revert(rtrans, newtable, rcache)
+end
+
+function _filter(name)
+  _keep(char) = isdigit(char) || isletter(char) || occursin(char, String(delim))
+  filter(char -> _keep(char), name)
+end
+
```
Comment (juliohm) [inline src/transforms/stdnames.jl:55]:
```suggestion
_clean(name) = filter(c -> isdigit(c) || isletter(c) || c ∈ delim, name)
```

File: src/transforms/stdnames.jl [hunk 42]
```diff
@@ -0,0 +1,76 @@
...
+function apply(transform::StdNames, table)  
+  cols = Tables.columns(table)
+  oldnames = string.(Tables.columnnames(cols))
+  preprocessed = _filter.(oldnames)
+  
+  spec = transform.spec
+
+  spec == :camel && (names = _camel.(preprocessed))
+  spec == :snake && (names = _snake.(preprocessed))
+  spec == :upper && (names = _upper.(preprocessed))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:35]:
```suggestion
  cleaned = _clean.(oldnames)
  
  spec = transform.spec

  spec == :camel && (names = _camel.(cleaned))
  spec == :snake && (names = _snake.(cleaned))
  spec == :upper && (names = _upper.(cleaned))
```

Comment (juliohm) [reply src/transforms/stdnames.jl:35]: Always try to choose short names for variables.

File: src/transforms/stdnames.jl [hunk 43]
```diff
@@ -0,0 +1,76 @@
...
+    while n ∈ newnames
+      n = string(n, "_")
+    end
+    push!(newnames, n)
+  end
+
+  newnames
+end
+
+const delim = [' ', '\t', '-', '_']
```
Comment (juliohm) [inline src/transforms/stdnames.jl:70]: Now that this delim variable is used in _clean, it needs to appear earlier in the file. Someone reading the code will try to find where it is defined. Think in terms of readers always.

**Review by juliohm — CHANGES_REQUESTED** [review]
I think we are almost done with this one @ceferisbarov , can you please take a look at this new set of revisions?

File: src/transforms/stdnames.jl [hunk 44]
```diff
@@ -0,0 +1,71 @@
...
+function apply(transform::StdNames, table)  
+  cols = Tables.columns(table)
+  oldnames = string.(Tables.columnnames(cols))
+  cleaned = _clean.(oldnames)
+  
+  spec = transform.spec
+
+  spec == :camel && (names = _camel.(cleaned))
+  spec == :snake && (names = _snake.(cleaned))
+  spec == :upper && (names = _upper.(cleaned))
```
Comment (juliohm) [inline src/transforms/stdnames.jl:34]:
```suggestion
  # retrieve spec
  spec = transform.spec
  
  # retrieve column names
  cols = Tables.columns(table)
  oldnames = string.(Tables.columnnames(cols))
  
  # clean column names
  cleaned = _clean.(oldnames)

  # apply spec
  spec == :camel && (names = _camel.(cleaned))
  spec == :snake && (names = _snake.(cleaned))
  spec == :upper && (names = _upper.(cleaned))
```

File: src/transforms/stdnames.jl [hunk 45]
```diff
@@ -0,0 +1,71 @@
...
+
+  newnames = _unique(names)
+  oldnew = Dict(zip(oldnames, newnames))
+
+  rtrans = Rename(oldnew)
+  newtable, rcache = apply(rtrans, table)
```
Comment (juliohm) [inline src/transforms/stdnames.jl:40]:
```suggestion
  # make names unique
  newnames = _unique(names)
  
  # rename transform
  rtrans = Rename(Dict(oldnames .=> newnames))
  newtable, rcache = apply(rtrans, table)
```

**Review by juliohm — CHANGES_REQUESTED** [review]

**Comment by juliohm** [comment]
@eliascarv anything else I missed before we can merge this?

**Review by eliascarv — APPROVED** [review]

**Comment by eliascarv** [comment]
For me everything is excellent.

**Comment by juliohm** [comment]
Another great addition @ceferisbarov ! 🎉
